### ISSUE TITLE
`.zip(None)` is always `None` so it should be linted

### ISSUE THREAD

**Issue by Veykril** [rust-lang/rust-clippy#17462]
### What it does

A new lint that checks for calls of the form `Option::zip` with `None` as the right hand side. This always yields `None` and is never going to be correct as one could just write `None`. https://github.com/rust-lang/rust-analyzer/pull/22924 is an example of a logic bug where the intent was to instead tuple the option with a `None`.

### Example

```rust
foo.zip(None)
```

Could be written as:

```rust
None
```
if intended, or `foo.map(|foo| (foo, None))` if the intent was completely different and thus a bug to be written that way.

**Comment by ChayimFriedman2** [comment]
This can also be applied to iterators.
